src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs

```diff
@@ -67,60 +64,91 @@ private bool TryInvokeSnippetCompletion(TabKeyCommandArgs args, CompletionServic
+            // i.e. the picker with all the folders and snippet organization.
+            //
+            // However, if the user instead has something like
+            //
+            //      "<line start><spaces><identifier><question><tab>"
+            //
+            // Then we take over and we show a completion list with all snippets in it
+            // in a flat list.  This enables simple browsing and filtering of all items
+            // based on what the user typed so far.
+            //
+            // If we detect this pattern, then we delete the previous character (the
+            // question mark) and we don't send the tab through to the editor.  In 
+            // essence, the <quesiton><tab> acts as the trigger, and we act as if that
+            // text never makes it into the buffer.
             Workspace workspace = null;
 
-            if (Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out workspace))
+            if (!Workspace.TryGetWorkspace(subjectBuffer.AsTextContainer(), out workspace))
             {
-                var documentId = workspace.GetDocumentIdInCurrentContext(subjectBuffer.AsTextContainer());
-                if (documentId != null)
-                {
-                    var document = workspace.CurrentSolution.GetDocument(documentId);
-                    if (document != null)
-                    {
-                        var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-
-                        if (caretPoint >= 2 && text[caretPoint - 1] == '?' && QuestionMarkIsPrecededByIdentifierAndWhitespace(text, caretPoint - 2, syntaxFacts))
-                        {
-                            var textChange = new TextChange(TextSpan.FromBounds(caretPoint - 1, caretPoint), string.Empty);
-                            workspace.ApplyTextChanges(documentId, textChange, CancellationToken.None);
-                            this.StartNewModelComputation(completionService, new CompletionTrigger(CompletionTriggerKind.Snippets), filterItems: false);
-                            return true;
-                        }
-                    }
-                }
+                return false;
+            }
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(subjectBuffer.AsTextContainer());
+            if (documentId == null)
+            {
+                return false;
+            }
+
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+            if (document == null)
+            {
+                return false;
+            }
+
+            var rules = GetCompletionService().GetRules();
+            if (rules.SnippetsRule != SnippetsRule.IncludeAfterTypingIdentifierQuestionTab)
+            {
+                return false;
             }
 
-            return false;
+            var syntaxFactsOpt = document.GetLanguageService<ISyntaxFactsService>();
```

Comment:
Yes.  Typescript has no ISyntaxFactsService.

Frankly, i'd like to extract out this piece of syntaxfacts as we often ask the following questions in several places:

1. Are you a case sensitive or insensitive language?
2. Is this an identifier in your language?

I think those should be in a public interface that TS could actually provide.  Right now expecting someone to implement all of ISyntaxFactsService is not suitable.